### PR TITLE
Deprecate using Scala 3 hard keywords as identifiers

### PR DESCRIPTION
**Pull request by smarter**

### CODE REVIEW

**Comment by dwijnand** [comment]
```
[error] /home/travis/build/scala/scala/src/reflect/scala/reflect/io/ZipArchive.scala:243:9: Wrap `enum` in backticks to use it as an identifier, it will become a keyword in Scala 3.
[error]     val enum    = zipFile.entries()
[error]         ^
[error] /home/travis/build/scala/scala/src/reflect/scala/reflect/io/ZipArchive.scala:246:14: Wrap `enum` in backticks to use it as an identifier, it will become a keyword in Scala 3.
[error]       while (enum.hasMoreElements) {
[error]              ^
[error] /home/travis/build/scala/scala/src/reflect/scala/reflect/io/ZipArchive.scala:247:24: Wrap `enum` in backticks to use it as an identifier, it will become a keyword in Scala 3.
[error]         val zipEntry = enum.nextElement
[error]                        ^
[error] /home/travis/build/scala/scala/src/reflect/scala/reflect/runtime/JavaMirrors.scala:198:26: Wrap `enum` in backticks to use it as an identifier, it will become a keyword in Scala 3.
[error]         def enumToSymbol(enum: Enum[_]): Symbol = {
[error]                          ^
[error] /home/travis/build/scala/scala/src/reflect/scala/reflect/runtime/JavaMirrors.scala:199:47: Wrap `enum` in backticks to use it as an identifier, it will become a keyword in Scala 3.
[error]           val staticPartOfEnum = classToScala(enum.getClass).companionSymbol
[error]                                               ^
[error] /home/travis/build/scala/scala/src/reflect/scala/reflect/runtime/JavaMirrors.scala:200:54: Wrap `enum` in backticks to use it as an identifier, it will become a keyword in Scala 3.
[error]           staticPartOfEnum.info.declaration(TermName(enum.name))
[error]                                                      ^
[error] 6 errors found
[error] (reflect / Compile / compileIncremental) Compilation failed
```

**Comment by dwijnand** [comment]
```
!!    1 - run/reflection-java-crtp                  [output differs]
% diff /home/jenkins/workspace/scala-2.13.x-validate-main/test/files/run/reflection-java-crtp.check /home/jenkins/workspace/scala-2.13.x-validate-main/test/files/run/reflection-java-crtp-run.log
@@ -1 +1,2 @@
+warning: 2 deprecations (since 2.13.7); re-run with -deprecation for details
 (type E,type E,true)
```

```
[error] /home/jenkins/workspace/scala-2.13.x-validate-main/test/junit/scala/collection/FactoriesTest.scala:218:12: Wrap `enum` in backticks to use it as an identifier, it will become a keyword in Scala 3.
[error]     object enum extends Enumeration {
[error]            ^
[error] /home/jenkins/workspace/scala-2.13.x-validate-main/test/junit/scala/collection/FactoriesTest.scala:222:22: Wrap `enum` in backticks to use it as an identifier, it will become a keyword in Scala 3.
[error]     val enumValues = enum.values
[error]                      ^
[error] two errors found
```

**Review by dwijnand — APPROVED** [review]

**Comment by SethTisue** [comment]
should this be guarded by `-Xsource:3`?

I guess this is a debatable gray area, but I guess I would have expected one of the two approaches:

* deprecate under `-Xsource:3`
* deprecate by default, and error under `-Xsource:3`

but this is neither which puzzles me a bit

**Comment by smarter** [comment]
> deprecate by default, and error under -Xsource:3

I could do that, but I don't think that gives much value-add, you'll get the error from cross-compiling with scala 3 anyway. The goal here is to give a heads-up to the people who are not already cross-compiling and do not use weird compiler flags :).

**Comment by SethTisue** [comment]
> The goal here is to give a heads-up to the people who are not already cross-compiling and do not use weird compiler flags

I would argue that such heads-up are merely annoying, since the issue in question is only relevant to Scala 3 migration and not relevant to anyone staying on Scala 2. In my view, `-Xsource:3` is the way we provide for people to opt-in to such heads-up-ing.

Deprecate-by-default is useful when something is actually on its way out of the language because it's problematic. But that doesn't apply to mere naming choices.

Curious what other folks think, though.

**Comment by dwijnand** [comment]
I'm a little concerned that enabling `-Xsource:3` might not always be a free choice.  https://github.com/scala/bug/issues/12437 is some evidence to that, but I might just be paranoid.  But, with that thinking, I think having it on by default is the right choice.  Anyone staying on Scala 2 can suppress it.

**Comment by smarter** [comment]
> In my view, -Xsource:3 is the way we provide for people to opt-in to such heads-up-ing.

I think the general way library/compiler changes go is: feature X is deprecated in version N, then removed in version N+k, no need to opt into flags. So anything that is removed in Scala 3 should be deprecated by default in Scala 2.13 (and I don't think catering to people who will never upgrade is useful?)

**Comment by SethTisue** [comment]
>  I don't think catering to people who will never upgrade is useful

(I wouldn't characterize it that way — I'm thinking of people who aren't in the process of upgrading _at the moment_, not people who will "never" upgrade.)

**Comment by dwijnand** [comment]
But that can apply to any deprecation warning..

File: src/reflect/scala/reflect/io/ZipArchive.scala [hunk 1]
```diff
@@ -240,11 +240,11 @@ final class FileZipArchive(file: JFile, release: Option[String]) extends ZipArch
     val root = new DirEntry(RootEntry)
     dirs.put(RootEntry, root)
     val zipFile = openZipFile()
-    val enum    = zipFile.entries()
+    val `enum`  = zipFile.entries()
```
Comment (NthPortal) [inline src/reflect/scala/reflect/io/ZipArchive.scala]: honestly, I think `entries` would be a better name, once we're already changing it

```suggestion
    val entries  = zipFile.entries()
```

**Comment by lrytz** [comment]
In general we try to deprecate things in 2.13 (without providing additional flags) that will be errors in Scala 3, and also make them errors in `-Xsource:3`. Procedure syntax is an example.

For the new keywords however, I'm a bit hesitant. When getting a warning about procedure syntax it's easy to fix once and be done. But if you're using a library that has `export` or `enum` in its API, you have no choice but keep writing backticks which can be quite annoying.

How about deprecating the identifiers only at definitions, but not usages?

**Comment by smarter** [comment]
Happy to do the change, but I'm not sure I get it: users will get errors at both definition and use-sites when they upgrade, so isn't it natural they get warnings for both on Scala 2.13? Especially since they can turn off whatever warnings they deem too annoying.

**Comment by lrytz** [comment]
The difference is, IMO, that it affects 2.13.x users at a "random" point in time (upgrade to the latest 2.13.x, not when starting to cross-build with 3), combined with the fact that they have to continuously work around / silence the warnings and cannot fix the issue at the root on their own, they have to wait for their dependencies to introduce a new name.

**Comment by smarter** [comment]
Ah so this gives time for library authors to come up with a new name, I see how that could be helpful indeed although it won't help with Java libraries.

**Comment by SethTisue** [comment]
Now that it's definition-site only, I'm fine with not putting it behind `-Xsource:3`.

**Comment by lrytz** [comment]
There are no warnings for bindings introduced in patterns (`val List(enum) = List(1)`, `case enum => ...`), but that's not so easy to do in the parser. So LGTM, @smarter let me know if I should merge it as-is.

**Comment by smarter** [comment]
Sounds good to me.

**Pull request merged**